Snowflake.Data/Core/FileTransfer/EncryptionProvider.cs

```diff
@@ -129,6 +129,95 @@ static private byte[] encryptFileKey(byte[] masterKey, byte[] unencryptedFileKey
+            string ivBase64 = encryptionMetadata.iv;
+
+            // Get decoded key from base64 encoded value
+            byte[] decodedMasterKey = Convert.FromBase64String(encryptionMaterial.queryStageMasterKey);
+
+            // Get key bytes and iv bytes from base64 encoded value
+            byte[] keyBytes = Convert.FromBase64String(keyBase64);
+            byte[] ivBytes = Convert.FromBase64String(ivBase64);
+
+            // Create temp file
+            string tempFileName = Path.Combine(Path.GetTempPath(), Path.GetFileName(inFile));
+
+            // Create decipher with file key, iv bytes, and AES CBC
+            byte[] decryptedFileKey = decryptFileKey(decodedMasterKey, keyBytes);
+
+            // Create key decipher with decoded key and AES ECB
+            byte[] decryptedBytes = CreateDecryptedBytes(
+                inFile,
+                decryptedFileKey,
+                ivBytes);
+
+            File.WriteAllBytes(tempFileName, decryptedBytes);
+
+            return tempFileName;
+        }
+
+        /// <summary>
+        /// Decrypt the newly generated file key using the master key.
+        /// </summary>
+        /// <param name="masterKey">The key to use for encryption.</param>
+        /// <param name="unencryptedFileKey">The file key to encrypt.</param>
+        /// <returns>The encrypted key.</returns>
+        static private byte[] decryptFileKey(byte[] masterKey, byte[] unencryptedFileKey)
+        {
+            Aes aes = Aes.Create();
+            aes.Key = masterKey;
+            aes.Mode = CipherMode.ECB;
+            aes.Padding = PaddingMode.PKCS7;
+
+            MemoryStream cipherStream = new MemoryStream();
+            CryptoStream cryptoStream = new CryptoStream(cipherStream, aes.CreateDecryptor(), CryptoStreamMode.Write);
+            cryptoStream.Write(unencryptedFileKey, 0, unencryptedFileKey.Length);
+            cryptoStream.FlushFinalBlock();
+
+            return cipherStream.ToArray();
+        }
+
+        /// <summary>
+        /// Creates a new byte array containing the decrypted data.
+        /// </summary>
+        /// <param name="inFile">The path of the file to write onto the stream.</param>
+        /// <param name="key">The encryption key.</param>
+        /// <param name="iv">The encryption IV or null if it needs to be generated.</param>
+        /// <returns>The encrypted bytes.</returns>
+        static private byte[] CreateDecryptedBytes(
+            string inFile,
+            byte[] key,
+            byte[] iv)
+        {
+            Aes aes = Aes.Create();
+            aes.Key = key;
+            aes.Mode = CipherMode.CBC;
+            aes.Padding = PaddingMode.PKCS7;
+            aes.IV = iv;
+
+            MemoryStream targetStream = new MemoryStream();
+            CryptoStream cryptoStream = new CryptoStream(targetStream, aes.CreateDecryptor(), CryptoStreamMode.Write);
+
+            byte[] inFileBytes = File.ReadAllBytes(inFile);
```

Comment:
ReadAllBytes() at once might not be not a good idea when it is a big file. Maybe using FileStream and ReadAsync when it is large (e.g. hundreds MB to dozens GB) could help for memory consumption and performance, please test with large files to make sure the performance is acceptable and comparable with snowsql.

Reply:
The same comment apply to CreateEncryptedBytes. I am ok if we handle large file and performance improvement later. But I would like to have some numbers to help PM and us to understand the performance handling large files and how it compare with with snowsql's PUT and GET. So we could estimate the priority of the performance improvement.